flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java

```diff
@@ -0,0 +1,902 @@
+
+        public StateFuture<TriggerResult> onElement(StreamRecord<IN> element) throws Exception {
+            return trigger.onElement(
+                    element.getValue(), element.getTimestamp(), window.get(), this);
+        }
+
+        public StateFuture<TriggerResult> onProcessingTime(long time) throws Exception {
+            return trigger.onProcessingTime(time, window.get(), this);
+        }
+
+        public StateFuture<TriggerResult> onEventTime(long time) throws Exception {
+            return trigger.onEventTime(time, window.get(), this);
+        }
+
+        public void onMerge(Collection<W> mergedWindows) throws Exception {
+            throw new UnsupportedOperationException("Merging window not support yet");
+        }
+
+        public StateFuture<Void> clear() throws Exception {
+            return trigger.clear(window.get(), this);
+        }
+
+        @Override
+        public String toString() {
+            return "Context{" + "key=" + key + ", window=" + window + '}';
+        }
+    }
+
+    /** Internal class for keeping track of in-flight timers. */
+    protected static class Timer<K, W extends Window> implements Comparable<Timer<K, W>> {
+        protected long timestamp;
+        protected K key;
+        protected W window;
+
+        public Timer(long timestamp, K key, W window) {
+            this.timestamp = timestamp;
+            this.key = key;
+            this.window = window;
+        }
+
+        @Override
+        public int compareTo(Timer<K, W> o) {
+            return Long.compare(this.timestamp, o.timestamp);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            Timer<?, ?> timer = (Timer<?, ?>) o;
+
+            return timestamp == timer.timestamp
+                    && key.equals(timer.key)
+                    && window.equals(timer.window);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = (int) (timestamp ^ (timestamp >>> 32));
+            result = 31 * result + key.hashCode();
+            result = 31 * result + window.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "Timer{"
+                    + "timestamp="
+                    + timestamp
+                    + ", key="
+                    + key
+                    + ", window="
+                    + window
+                    + '}';
+        }
+    }
+
+    protected <S extends State, N, T> S getAsyncPartitionedState(
+            N namespace, TypeSerializer<N> namespaceSerializer, StateDescriptor<T> stateDescriptor)
+            throws Exception {
+        return getAsyncKeyedStateBackend()
+                .getOrCreateKeyedState(namespace, namespaceSerializer, stateDescriptor);
+    }
```

Comment:
I'd suggest this method includes `stateWrapWithNamespace`. WDYT?

And no need for `Async`, just `getPartitionedState ` seems good

Reply:
Resolved, I just delete this useless function and let the caller directly use ` getAsyncKeyedStateBackend().getOrCreateKeyedState()`, See

```
Context#getPartitionedState(StateDescriptor<T> stateDescriptor)
```